### ISSUE TITLE
Wandering monks

### ISSUE THREAD

**Issue by matsen** [matsengrp/pt#11]
@thayerf -- @bcclaywell and I were talking yesterday about how the duplication of Partition objects has been making pain for us in a variety of ways, and that perhaps we should consider an alternate strategy in which there is a fixed number of non-reproducing threads (that we dubbed the "wandering monks") that do the exploration rather than the currently uber-fecund thread setup. 

Each one of these threads would have a Partition object that gets changed in-place (e.g. NNI moves) rather than upon duplication. We view this as an advantage, because it could lead to better encapsulation of the data, and assurance that the tree and the partition data are in sync.

To implement this we are going to need to figure out a few things.

* What will a thread do when it exhausts part of the search tree? It seemed to us that if we record the steps from where we started, then we can go back to where we were before and start wandering anew.
* However, monks may get "trapped" if they explore a part of the search tree and then when they go back to their original start point, the other neighbor trees have been visited by another monk. Thus there should be a means for them to get teleported or reincarnated in another, open, part of the search space. 
* Similarly, we need a way to start all of the threads heading somewhat in different directions. 
* So, it seems like we should either have a way for still-active monks to reproduce to split up the work of exploring open space, or for them to get transported to an open part of the space. The same strategy could be used for both resolving trapped monks and to populate the graph with as many monks as desired.

Thoughts?

**Comment by thayerf** [comment]
This sounds like a great idea! The only thing I'm concerned about is how the monks will exhaust all neighbors. For example, if the monk finds 3 neighbors of its starting point, how would it explore all 3 without recalculating the clv's each time?

**Comment by cwhidden** [comment]
This seems like a cool idea, but I'm not sure that it's actually any simpler. When you return to the previous tree you will need to initialize the Partition object for that tree. This will require either a new method to initialize an appropriate partition for a given topology (which is what Thayer was trying to avoid doing) or copying the Partitions anyway. If you do the former, then I think it would be more efficient to simply maintain a list of "open" trees to search and have each monk explore the next tree from this list. That was my original idea of how to do this but 1. Thayer's idea of storing the whole Partition was easier to implement and 2. I'm not sure if it actually is any faster to do so.

**Comment by cwhidden** [comment]
And Thayer just said that more succinctly while I was interrupted :)

**Comment by bcclaywell** [comment]
Assuming I'm understanding everything correctly, the tree we came from is connected to the current tree by a single NNI move, so we can keep a stack of the NNI moves used to arrive at the current tree. Any time we need to go back a step (e.g., we've reached a dead end), we pop the most recent move off the stack and invalidate just the CLVs on either side of the edge in question. Then we can do a partial traversal to resynchronize the tree and the partition. We can do this indefinitely if we need to continue going backward, one step at a time. This saves us the trouble of reinitializing or copying the partition object.

**Comment by matsen** [comment]
Responding to @thayerf , it will do some recalculation, but just locally around the NNI move. This is cheap.

The motivation for doing this is that we are currently encountering problems with out-of-control memory usage. We can't store every possible place we want to explore-- memory blows up. @bcclaywell was thinking that he could delay copying of the partition objects, but it turns out that this is going to require a lot more recalculation than required by one NNI swap. Right now it's doing a full traversal, which is O(n) more expensive than one NNI swap.

And to make sure it's clear, responding to @cwhidden , 

> When you return to the previous tree you will need to initialize the Partition object for that tree

we are proposing modifying the tree and partition in place as @bcclaywell described. I should also note that this backtracking recomputation happens less frequently than forward exploration.

**Comment by bcclaywell** [comment]
As for how to handle monks getting trapped or going idle when they exhaust their part of the search space, I envision there being some celestial authority keeping track of which monks are active and which are idle. When an active monk is preparing to make a move to a new tree, he first asks the authority if there are any idle monks, and if so, teleports one of them to the new tree instead and moves onto the next proposal. Otherwise, he makes the move himself.

I think a teleported monk would have to make a full traversal of the tree one he arrives to synchronize everything, but beyond that would proceed as any other.

**Comment by cwhidden** [comment]
Okay, that sounds fine then if only a small part of the partition has to change when you backtrack. What happens when you backtrack though? Do you now re-optimize the weight of the modified edge and then explore the neighbors of that tree again, relying on some data structure to avoid revisiting trees? If you are only maintaining a linear-size path for each monk then you can afford to remember every open tree along this path.

**Comment by cwhidden** [comment]
In fact, that suggests to me that the problem with the current approach is that the search is arbitrary, so that you are opening up too many trees and not fully exploring any neighborhoods. If it was a depth first search then that could not happen because every monk is at most nlogn moves from the start point and you only have open trees along these paths. Is there any way to tweak the order of exploration?

Note that I said a linear path above, but it is nlogn for NNI, I was thinking of SPR.

**Comment by cwhidden** [comment]
Otherwise, Brian's strategy sounds ideal for teleporting. You could potentially use some sort of criteria to choose neighbors in a dense area like the highest average percentage of high likelihood neighbors as well.

**Comment by bcclaywell** [comment]
Hm, yeah, branch length optimization does throw a wrench into things. We end up having to do a full traversal at every step anyway, backwards or forwards, right? Even if we cached the optimized branch lengths as part of the move rollback data, we'd have to traverse the tree again to update all the CLVs.

**Comment by matsen** [comment]
Given this, do you think that the delayed-allocation and full traversal strategy is viable, @bcclaywell ?

Is the allocation and copying overhead anything compared to that of a full traversal of a big tree with lots of columns?

**Comment by bcclaywell** [comment]
It might still be an option, yeah, but the question then becomes, why are we copying any inner node data at all if we know we have to do a full traversal anyway to optimize the branch lengths after every move? That makes the wandering monk strategy a little more attractive from the standpoint of only ever having one partition object per thread to worry about, and the tip node data on that partition will already be set for us.

In thinking about it a little bit more, with the monks I don't think we'd have to optimize the tree again after a step backwards, since by that time we will have already evaluated the log-likelihood of that tree and really only care about the topology for generating the next move. After that, we make the next NNI move and have to do branch length optimization again, requiring another full traversal.

**Comment by cwhidden** [comment]
Maybe I'm misremembering, but don't we only do a full traversal once for each good tree right now and only a partial traversal for each considered neighbor? It sounds like you are talking about doing a full traversal for bad trees as well.

**Comment by bcclaywell** [comment]
And maybe I'm misunderstanding, but before we can decide if a tree is good or bad, we have to do [branch length optimization on the entire tree](https://github.com/matsengrp/pt/blob/63e9391dd9e817092925db78170ea21955382824/src/partition.cpp#L513) before we compute its log-likelihood, which equates to doing a full traversal, right?

**Comment by cwhidden** [comment]
Again, it's been a while but I though we were only optimizing the length of the new branch to get something half-decent and using that as the criteria for which trees are good. We then do full branch length optimization when we actually explore a good tree.

**Comment by cwhidden** [comment]
@thayerf can you clarify the traversal choice and branch length optimization process?

**Comment by bcclaywell** [comment]
Ah, I see. I'm just going by what's written, so I'll defer to you, @thayerf, and @matsen for intent.

**Comment by cwhidden** [comment]
Yeah, that was the ideal but maybe we couldn't get it to work. I've pinged Thayer for clarification.

Again, I don't know which method would actually be faster, I just want to make sure we consider all of the tradeoffs. I think we can probably also do something that is a hybrid of the two ideas by giving the monks a limited amount of memory.

**Comment by thayerf** [comment]
```cpp
if (all.insert(label, 0.0)) {
    FullBranchOpt(clone);
    /// @todo Can we save time by having FullBranchOpt return this?
    double lambda_1 = FullTraversalLogLikelihood(clone);
    // Compare new likelihood to ML, then decide which table to put in.
```
So it optimizes all branch lengths for every tree right now. The time saved from partial traversals comes from not having to update all the CLV's after NNI, and from doing partial traversals to update CLV's during the branch length optimization process.  So there's a full traversal for each new tree right now. If we wanted to be greedier we could have it just optimize the NNI branch after swapping, check the likelihood, and then only optimize the whole tree if it meets the likelihood threshold.

**Comment by cwhidden** [comment]
Okay, then that was just something we discussed but didn't have time for or couldn't get working easily. Thanks!

**Comment by matsen** [comment]
Tell me if I'm off the mark here, but it seems to me that the only viable means of continuing the current strategy (with delayed allocation) is to do a full traversal on the tree. [For those following from home, this was a discussion that @bcclaywell and I had yesterday. It's a little hard to explain in an issue, so perhaps you can believe us until an IRL meeting?] So, if we are interested in doing a strategy in which we only partially optimize branch lengths, it seems like monks are the way to go. However, I note that it will require some clever data structures to keep track of which branch lengths have been modified on the monk's path if we want to efficiently backtrack. I wonder if we could do this with some vector accumulated from a traversal. When we backtrack we could make sure that we are "dirtying" the proper nodes for our partial traversal.

Responding to @cwhidden : 

> Do you now re-optimize the weight of the modified edge and then explore the neighbors of that tree again, relying on some data structure to avoid revisiting trees? If you are only maintaining an nlogn path for each monk then you can afford to remember every open tree along this path.

Not sure if I totally understand this question, but the tree revisiting-avoidance is done by the hashtable, right? 

> In fact, that suggests to me that the problem with the current approach is that the search is arbitrary, so that you are opening up too many trees and not fully exploring any neighborhoods. If it was a depth first search then that could not happen because every monk is at most nlogn moves from the start point and you only have open trees along these paths. Is there any way to tweak the order of exploration?

I don't see this, sorry! It seems like we are indeed doing a DFS, which stops at a given level when it runs out of good trees and backs up to the previous level. This is an essential point, though, so let me know if I'm thinking wrongly.

**Comment by cwhidden** [comment]
Yes, the tree revisiting-avoidance is done by the hashtable. My first question was whether the monks would have to check whether neighbors of a given tree are in the hashtable multiple times. So suppose you start at a tree T. You then visit one of its neighbors T_1 and add it to the hashtable. You explore T_1's neighbors and eventually go back to T. You then look at T_1, see that it is in the hashtable and move on to T_2. Repeating this, you lookup T's neighbors 1 + 2 + 3 + ... x = O(x^2) times, for x neighbors (O(n) for NNI). You might not have to do this for all of them (e.g. if you visit a neighbor from another tree / monk) but it is quadratic in the number of actually visited neighbors.

I probably wrote that paragraph strangely. With the monks you would indeed be doing a DFS. I was talking about the current implementation which sends a thread to any arbitrary neighbor. There is no guarantee that the paths taken by the active threads are on a linear path from the starting tree, and I would guess that they visit the oldest trees in order. When I said "monk" here I really meant thread.

Any monk is O(nlogn) NNI operations away from the start tree, so there should be no problem maintaining a list of modified branch lengths. I was also suggesting that you could potentially even keep a list of neighbors that still need to be visited from each of the trees on that path, for a total of O(n^3logn) storage (On the order of nlogn path, n neighbors, n-length strings). That would avoid the extra hashtable queries.

**Comment by bcclaywell** [comment]
Yes, I anticipated storing a list of the moves available to a monk upon arrival at a new tree, so that when the monk pops back up the stack to an already-visited tree, the next move is available directly instead of having to go through all the neighbors again to see if they're in the hash table.

I'm having trouble following your storage complexity requirement though. Each of the n neighbors of the current tree corresponds to a move in the list mentioned above, and each move can be encoded as a pointer to an inner edge and an NNI "direction" (in libpll, "right" or "left"). At the monk's current search depth m, this would require O(mn) storage, since the monk needs to keep track of the moves still available for each of the m trees on its current path from the starting tree.

**Comment by cwhidden** [comment]
@bcclaywell The depth m is O(nlogn) for NNI. I was thinking that you would need O(n) storage for each neighbor but you are right that you can do this with constant storage, so that is a total of just O(n^2 logn) for each monk.

**Comment by bcclaywell** [comment]
Oh, I get it! Sorry, I was thinking in implementation mode and wondering how I was going to have an array of length nlogn. :)

**Comment by matsen** [comment]
Any consensus here, folks?

I don't want to pollute our thinking, but for another project (https://github.com/matsengrp/linearham) it would be actually really nice to have a self-contained object with all the libpll information and methods to change around the tree, which is exactly what a monk would be. So if we're on the fence that might be a tie-breaker.
